aten/src/ATen/native/mobile/cpu/internal/Add.cpp

```diff
@@ -0,0 +1,244 @@
+#include <ATen/native/mobile/cpu/Engine.h>
+
+#ifdef USE_XNNPACK
+
+#include <ATen/native/mobile/cpu/internal/Allocator.h>
+#include <ATen/native/mobile/cpu/internal/Common.h>
+
+namespace at {
+namespace native {
+namespace mobile {
+namespace cpu {
+namespace internal {
+namespace add {
+
+struct Context final {
+  Operator add_op;
+
+  static constexpr float kMin = -std::numeric_limits<float>::infinity();
+  static constexpr float kMax = std::numeric_limits<float>::infinity();
+};
+
+namespace {
+
+// Supports NHWC and NCHW FP32 adds.
+
+bool available(
+    const int64_t input1_channels,
+    const int64_t input2_channels,
+    const int64_t output_channels,
+    const Scalar output_min,
+    const Scalar output_max) {
+         // Mobile
+  return mobile::cpu::available() &&
+         // Input1
+         (input1_channels > 0) &&
+         // Input2
+         (input2_channels > 0) &&
+         // Input1 == Input2
+         (input1_channels == input2_channels) &&
+         // Output
+         (output_channels > 0) &&
+         // Input == Output
+         (input1_channels == output_channels) &&
+         // Output Min / Max
+         (output_min.isIntegral(true) || output_min.isFloatingPoint()) &&
+         (output_max.isIntegral(true) || output_max.isFloatingPoint()) &&
+         (output_max.to<float>() > output_min.to<float>());
+}
+
+Context create(
+    const int64_t input1_channels,
+    const int64_t input2_channels,
+    const int64_t output_channels,
+    const Scalar output_min,
+    const Scalar output_max) {
+  TORCH_INTERNAL_ASSERT(
```

Comment:
I just want to confirm: when these asserts fail, they will tell the user, "Please report a bug in PyTorch". Do you want them to report a bug if this assert fails?

Reply:
No, that wasn't my intention.  Thanks for pointing that out.  This is supposed to be a sanity check.  How do you suggest I do that?

Reply:
Umm, usually a sanity check means that if it fails, there's a bug :) If the error condition is something that an end user could trigger by misusing your API, you should use `TORCH_CHECK` instead. (The line gets blurry when you create an API that is used by other developers, but generally speaking, use internal asserts for APIs consumed by people who regularly work on pytorch, and check for people who don't regularly work on pytorch.)